File: src/commands/count/postgrads.rs
use crate::prelude::*;

/// Count the number of registered postgraduate students.
#[poise::command(slash_command, guild_only)]
pub async fn postgrads(context: BotContext<'_>) -> BotResult<()> {
    // Defer reply to allow time for counting students
    context.defer().await?;

    // Count the total and registered postgrads
    let count = sqlx::query!(
        "select count(registered_user_id) as registered, count(*) as total
        from users
        where kind = ?",
        VerifiedUserKind::Postgrad.to_string(),
    )
    .fetch_one(&context.data().database_pool)
    .await?;

    // Reply to the command
    context
        .send(|reply| {
            reply.embed(embeds::registration::count(
                "Postgraduates",
                count.registered as u64,
                count.total as u64,
            ))
        })
        .await?;

    Ok(())
}

File: norris/commands/count/postgrads.py
from discord import ApplicationContext
from sqlalchemy import func, select

from ...bot import Norris
from ...model import VerifiedUser, VerifiedUserKind


async def handle_count_postgrads(norris: Norris, context: ApplicationContext) -> None:
    """
    Handles the `count postgrads` slash command.
    """
    # Defer response to allow time for query
    await context.response.defer()

    # Count the total and registered postgrads
    async with norris.database_engine.begin() as connection:
        result = await connection.execute(
            select(func.count(VerifiedUser.registered_user_id), func.count())
            .where(VerifiedUser.kind == VerifiedUserKind.POSTGRAD),
        )
        registered, total = result.one()

    from ...responses import embeds

    # Display results to user
    await context.followup.send(
        embed=embeds.registration.count("Postgraduates", registered, total),
    )
